tox + travis-ci changes

steenzout opened the pull request:
- cache directories where virtualenvs reside to speed up process
- trigger different builds per Python version
- massive shrink of `tox.ini`
- minor cleanup of `.travis.yml`

checked carefully and the number of virtualenvs tested on travis-ci is the same.

before changes [total elapsed time = 20m](https://travis-ci.org/pytest-dev/pytest-django/builds/156458481).

[currently it's around 15m](https://travis-ci.org/steenzout/python-pytest-django/builds/158359954).

steenzout commented:
anyone?

blueyed commented on .travis.yml:
```diff
@@ -2,31 +2,29 @@
+cache:
+  directories:
+    - "${HOME}/virtualenv"
+    - "${TRAVIS_BUILD_DIR}/.tox"
```
What about caching just `~/.cache/pip` instead?
After all the cache size is also an issue, since it will be fetched before the builds.

What is `${HOME}/virtualenv` used for?  Isn't that Travis' location for the Python?!

steenzout replied:
I think the packages on the `install:` section get installed on the `virtualenv`:

```
install:
  (...)
  - pip install tox==2.3.1
  - pip install tox-travis
```

so I typically add these but I can also remove this if you like.

steenzout replied:
I can try and test just caching ``/.cache/pip`.
I'll need to do it tonight though (I'm on US/Mountain Time).

if you want to merge this one and a new PR for the cache change just let me know.

steenzout replied:
FYI caching `.pip` is not good enough.
some files seem to be updated in the installation process therefore you'll always extra overhead on top of each build since the cache will always be changed and an upload will need to be made.

blueyed commented on tox.ini:
```diff
@@ -1,1978 +1,58 @@
+    - pypy3-pytest300-django1.8-{sqlite,sqlite_file}
+    - pypy3-pytest292-django1.8-sqlite_file
+    - pypy-pytest300-django1.10-sqlite_file
+    - py{35,34,33,27,py3,py}-checkqa
```
This (checkqa) was done for py27 and py35 only before.

steenzout commented:
sorry, I didn't see `generate_configurations.py`.
it crossed my mind you might have been using a script to maintain this big `tox.ini` though :)
I've removed it now.

for me it has been easy to maintain the envs between `tox.ini` and `.travis.yml`.

regarding the `--skip-missing-interpreters` I think the docs say it will return success if the interpreter is missing.
do you want me to add this and remove the `allow_failures`?

blueyed commented:
I think `allow_failures` is different.
It seems like on Travis with Python 3.5 you have all older Python versions, too (but not the other way around).  But this PR adds https://pypi.python.org/pypi/tox-travis, which seems to handle this better.

It's :+1: from me, although I somehow dislike that there is now only a build per Python version, which makes it hard to see at first glance that e.g. all django 1.10 builds failed (because you need to look at each job's output to figure that out).

@pelme 
What do you think?

steenzout commented:
I can make it run a single complete test per test environment in parallel.
it's pretty simple.

let me know.

blueyed commented:
> I can make it run a single complete test per test environment in parallel.

Sounds good.. let's see how it compares in term of duration.

steenzout commented:
FYI caching `.pip` is not good enough.
some files seem to be updated in the installation process therefore you'll always extra overhead on top of each build.

steenzout commented:
I've enabled the parallel builds, removed `tox-travis` due to that and changed the directories being cached (improves build performance and shortens cache size).

steenzout commented:
ping.

blueyed commented:
Seems to be much slower now in total ( 51 min 48 sec) vs initially (39 min 10 sec).
Elapsed time is 13 min 18/15 min 6 sec vs 13 min 33 sec now.

I am not really convinced that it's better than the previous approach through `generate-configurations.py`.

steenzout commented:
trigger a rebuild and see if the cache helps.

[mine](https://travis-ci.org/steenzout/python-pytest-django) was faster.

pelme commented:
Thanks a lot! This looks really good. I'm very happy to move to this. When `generate_configurations.py` was conceived, tox parametrization was not a thing. :)

pelme commented:
@steenzout I updated to pytest 3.0.3 and merged this. Thanks a lot @steenzout ! 👍 😃
